Add docs about sqlite labels (#919)
fermyon/developer: content/cloud/linking-applications-to-resources-using-labels.md

To facilitate the seamless transition of applications between different environments, such as from a local development environment to Fermyon Cloud, Spin applications don't reference databases by their physical location, like a connection string. Instead, they use a label. A label is an abstract identifier, such as "default," "user-accounts," or "transactions." The environment determines how this label is mapped to a physical database.

In the Spin development environment, that mapping is handled by the [runtime configuration file](/spin/dynamic-configuration#runtime-configuration).  In Fermyon Cloud, it's managed through _links_.

> Link management is currently available only for NoOps SQL databases. Fermyon Cloud currently allows applications only the "default" key-value store, which is private to the application and shares the application's lifespan. Key-value link management is planned for a future release of Fermyon Cloud.

> This may sound abstract, so let's walk through a specific example:

Reviewer: Already had the colon @itowlson so it was meant to be :)

> 

Reviewer: ```suggestion
```

> This is where Fermyon Cloud **links** and **labels** are helpful. As a Spin application developer, you provide a label whenever you fill out the name for a NoOps SQL database in the component manifest. Let's walk through a specific example together:

Reviewer: Agreed, prior to this paragraph, we mentioned links and labels 3 times each. I will remove it.

> This is where Fermyon Cloud **links** and **labels** are helpful. As a Spin application developer, you provide a label whenever you fill out the name for a NoOps SQL database in the component manifest. Let's walk through a specific example together:

Reviewer: ```suggestion
```

```toml
// --snip--
name = "todo-app"
[component]
id = "todo"
sqlite_databases = ["data"]
// --snip--
```

The string "data" is the **label** that the Spin application uses to reference the NoOps SQL Database. At application deployment time, we have to make a decision about how to map that label to an actual database.  The `spin cloud deploy` command gives us a couple of choices:
- Use an existing database and link app to it
- Create a new database and link the app to it

> In this example below, we are going to **link** the app to a specific NoOps SQL Database instance using the label. To do this, we choose the first option (that `spin deploy` gives us). We make our selection by using the up/down arrow keys and then `return` to continue:

Reviewer 2: ```suggestion
In this example below, we are going to **link** the app to an existing NoOps SQL Database instance using the label. To do this, we choose the first option (that `spin deploy` gives us). We make our selection by using the up/down arrow keys and then `return` to continue:
```

It always links to a "specific" database - "in this example" implies we are talking about the decision being made in the example, which us "link to existing" vs "create and link to new."

```bash
spin deploy
Uploading todo-app version 0.1.0-r234fe5a4 to Fermyon Cloud...
Deploying...
App "todo-app" accesses a database labeled "data"
Would you like to link an existing database or create a new database?:
> Use an existing database and link app to it
  Create a new database and link the app to it
```

Since we chose to link to an existing database, `spin deploy` asks us which database we would like to link this app's "data" label to:

```bash
Which database would you like to link to todo-app using the label "data":
  finance
> projects
```

> > Note: The database names are arbitrary and created on our behalf. In this case `inspiring-otter` is our Cloud database. We are linking the `inspiring-otter` database to our `todo-app` using the label "data".

Reviewer 2: ```suggestion
> Note: When Fermyon Cloud creates a new databases, it gives it an arbitrary name such as `inspiring-otter`. You can rename databases to give them more memorable names. In the example above, our databases have been renamed for `finance` and `projects`.
```

From this point, we see that our app is deploying and once finished we receive our deployed application's URL endpoint:

```bash
Waiting for application to become ready...... ready
Available Routes:
  todo-app: https://todo-app-ylnmlqmr.fermyon.app (wildcard)
```

Let's kick things up a notch and deploy a second Spin application (`todo-app-v2`) that will also use the same Cloud database:

```toml
// --snip--
name = "todo-app-v2"
[component]
id = "todo"
sqlite_databases = ["default"]
// --snip--
```

When deploying the second Spin application (`todo-app-v2`) we are again asked if we would like to use an existing database and link our application (in this case `todo-app-v2`) to it:

```bash
spin deploy       
Uploading todo-app-v2 version 0.1.0-r6e577b18 to Fermyon Cloud...
Deploying...
App "todo-app-v2" accesses a database labeled "default"
Would you like to link an existing database or create a new database?:
> Use an existing database and link app to it
  Create a new database and link the app to it
```

We again use the up/down arrow keys to make our selection and hit `return` to continue:

```bash
Which database would you like to link to todo-app-v2 using the label "default":

> > inspiring-otter

Reviewer: ```suggestion
  finance
> projects
```

```

> In the example here we have chosen to link `todo-app-v2` to our pre-existing `inspiring-otter` Cloud database using via the "default" label, that we added to our `todo-app-v2` application manifest (`spin.toml`) file.

Reviewer: ```suggestion
In the example here we have chosen to link `todo-app-v2` to our pre-existing `projects` Cloud database using via the "default" label, that we added to our `todo-app-v2` application manifest (`spin.toml`) file.
```

Now we have two separate Spin applications (`todo-app` and `todo-app-v2`) using their own labels to link to the same database. 

> Links and labels are application scoped. Whenever a specific Spin application (say `todo-app`) needs to access the database resource (`inspiring-otter`), it will do so via the label "data".

Reviewer: ```suggestion
Links and labels are application scoped. Whenever a specific Spin application (say `todo-app`) needs to access the database resource (`projects`), it will do so via the label "data".
```

> The Spin application does not need to know the true name of the Cloud resource (`inspiring-otter` database) that it is interacting with. 

Reviewer: ```suggestion
The Spin application does not need to know the true name of the Cloud resource (`projects` database) that it is interacting with. 
```

## Features of Links and Labels

The benefits of using labels and Fermyon Cloud links include:

* **Easy Sharing**: Share your Fermyon Cloud resource across applications effortlessly.
* **Resource Creation Control**: You decide when to create and delete Fermyon Cloud resources.
* **Decoupled Resource Lifecycle**: Allow Fermyon Cloud resources to have lifecycles independent of your Spin application's lifecycle.
* **Seamless Cloud Integration**: Cloud experience smoothly integrates with local Spin development, no runtime configuration changes required. 
* **Dynamic App Resource Configuration**: Change the data resource your application points to without rebuilding it.

## Working With Links and Labels

Whenever you add a string to your NoOps SQL database entry in the component manifest, you're creating a label. When you deploy your Spin application to Fermyon Cloud, you will be prompted to link the label to a specific NoOps SQL database instance. Later on, if you'd like to update the link you can do so via the `spin cloud link` command. 

> 

Reviewer 3: > Later on, if you'd like to update the link you can do so via the `spin cloud link` command.

We need to merge https://github.com/fermyon/cloud-plugin/pull/123 before the release to make this line true. cc @vdice

Reviewer: Thanks @kate-goldenring 
#123 seems to be merged so we are good to go with this line.

To see your existing labels, run the `spin cloud <resource-type> list` command, where `<resource-type>` is the subcommand for the resource whose links you want to see. For example, if you want to see how your applications and your NoOps SQL databases are linked, run the following command:

```bash
$ spin cloud sqlite list
+--------------------------------------------------------+
| App                          Label     Database        |
+========================================================+

> | todo-app                     data      inspiring-otter |

Reviewer: ```suggestion
| todo-app                     data      projects        |
```

> | todo-app-v2                  default   inspiring-otter |

Reviewer: ```suggestion
| todo-app-v2                  default   projects        |
```

+--------------------------------------------------------+
```

Now you can see your Spin applications, their respective labels and their connected NoOps SQL Databases.

> If you'd like to change your link while your Spin application is running, you can do so with the following command:

Reviewer 3: ```suggestion
Though not recommended, if you'd like to unlink your resource while your Spin application is running, you can do so with the following command:
```

```bash
spin cloud sqlite unlink --app todo-app data
```

> Now we've successfully unlinked our `todo-app` from the `inspiring-otter` database. 

Reviewer: ```suggestion
Now we've successfully unlinked our `todo-app` from the `projects` database. 
```
